main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala

```diff
@@ -0,0 +1,716 @@
+    case (Kind.Eff, Kind.Eff) =>
+      // first simplify the types to get rid of assocs if we can
+      for {
+        (t1, p1) <- simplifyType(tpe1, renv, eqEnv, loc)
+        (t2, p2) <- simplifyType(tpe2, renv, eqEnv, loc)
+        res0 <- EffUnification.unify(t1, t2, renv).mapErr(toTypeError(_, prov))
+        // If eff unification has any constrs in output, then we know it failed so the subst is empty
+        res =
+          if (res0._2.isEmpty) {
+            ResolutionResult.newSubst(res0._1)
+          } else {
+            ResolutionResult.constraints(List(TypingConstraint.Equality(t1, t2, prov)), p1 || p2)
+          }
+      } yield res
+
+    case (Kind.Bool, Kind.Bool) =>
+      // first simplify the types to get rid of assocs if we can
+      for {
+        (t1, p1) <- simplifyType(tpe1, renv, eqEnv, loc)
+        (t2, p2) <- simplifyType(tpe2, renv, eqEnv, loc)
+        res0 <- BoolUnification.unify(t1, t2, renv).mapErr(toTypeError(_, prov))
+        // If bool unification has any constrs in output, then we know it failed so the subst is empty
+        res =
+          if (res0._2.isEmpty) {
+            ResolutionResult.newSubst(res0._1)
+          } else {
+            ResolutionResult.constraints(List(TypingConstraint.Equality(t1, t2, prov)), p1 || p2)
+          }
+      } yield res
+
+
+    case (Kind.RecordRow, Kind.RecordRow) =>
+      // first simplify the types to get rid of assocs if we can
+      for {
+        (t1, _) <- simplifyType(tpe1, renv, eqEnv, loc)
+        (t2, _) <- simplifyType(tpe2, renv, eqEnv, loc)
+        res0 <- RecordUnification.unifyRows(t1, t2, renv).mapErr(toTypeError(_, prov))
+        // If eff unification has any constrs in output, then we know it failed so the subst is empty
+        res = if (res0._2.isEmpty) ResolutionResult.newSubst(res0._1) else throw InternalCompilerException("can't handle complex record stuff", SourceLocation.Unknown)
+      } yield res
+
+    case (Kind.SchemaRow, Kind.SchemaRow) =>
+      // first simplify the types to get rid of assocs if we can
+      for {
+        (t1, _) <- simplifyType(tpe1, renv, eqEnv, loc)
+        (t2, _) <- simplifyType(tpe2, renv, eqEnv, loc)
+        res <- SchemaUnification.unifyRows(t1, t2, renv).mapErr(toTypeError(_, prov))
+      } yield ResolutionResult.newSubst(res)
+
+    case (Kind.CaseSet(sym1), Kind.CaseSet(sym2)) if sym1 == sym2 =>
+      for {
+        (t1, _) <- simplifyType(tpe1, renv, eqEnv, loc)
+        (t2, _) <- simplifyType(tpe2, renv, eqEnv, loc)
+        res <- CaseSetUnification.unify(t1, t2, renv, sym1.universe, sym1).mapErr(toTypeError(_, prov))
+      } yield ResolutionResult.newSubst(res)
+
+    case (k1, k2) if KindUnification.unifiesWith(k1, k2) => resolveEqualityStar(tpe1, tpe2, prov, renv, eqEnv, loc)
+
+    case _ => Err(toTypeError(UnificationError.MismatchedTypes(tpe1, tpe2), prov))
```

Comment:
Mismatched kinds?

Reply:
Doesn't exist.

Added this to followup